### DISCUSSION TITLE
[General] Account Abstraction IAccount interface FAQ

### DISCUSSION THREAD

**Discussion by andrewpixelgemium** [zkSync-Community-Hub/zksync-developers#163] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to extend the IAccount interface to create a Smart Account contract for session key-pair management. Where session key-pair will represent an EOA that signs transactions

### What is the specific issue or error you're encountering?

Issue:
When a smart account performs a transaction, is the validation of the transaction (signature) conducted anywhere other than the methods within the IAccount interface? If so, where does it occur?

In the current scenario, we have created a session key-pair (Wallet.createRandom) and allowed the use of the session’s private key to sign a transaction by recording the session address in the Smart Account contract. 

In the validateTransaction method of the Smart Account Contract, we have allowed the transaction to proceed if the recovered address from the signature matches the session address. However, when executing a transaction with "from" set to the owner's address and signed with the session, we encounter the error "Account validation returned invalid magic value."

I have checked the transaction execution flow and it did not observe the methods from the IAccount interface within the Smart Account contract being executed.

### Can you share the error messages or logs you're receiving, if any?

reason: Validation revert: Account validation error: Account validation returned invalid magic value

### Have you made any recent changes to the contract before encountering this issue?

I'm success with owner as signer. But with session as signer, it's failed

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

The issue is occurring in Zksync system contract for sure. Because, all the methods in IAccount interface was not executed.

### Have you tried to isolate the problem, and if so, what were the results?

I figured, there must be somewhere where signature validation is performed other than the methods in the IAccount interface

### What steps have you already taken to try to resolve the issue?

Not yet

### Repo Link (Optional)

https://github.com/andrewpixelgoal3/g3-zg-contract.git

**Comment by andrewpixelgemium** [comment]
Any help?

**Comment by idea404** [comment, marked as answer]
Hi @andrewpixelgoal3 is this the contract you want to deploy as your custom account that uses session keys? 

> https://github.com/andrewpixelgoal3/g3-zg-contract/blob/main/contracts/Account.sol

I don't seem to catch any method that adds the session key to the authorised keyset (also no data collection similar to that here). I think you might be looking for a similar implementation to this:

```solidity
    ...
    
    KeyPair[MAX_KEY_PAIRS] private keyPairs;
     
    function addKeyPair(address _owner) public onlyValidKeyPair(msg.sender) {
        require(!keyPairsContain(_owner), "Key pair already added");
        require(numKeyPairs < MAX_KEY_PAIRS, "Maximum key pairs limit reached");
        keyPairs[numKeyPairs] = KeyPair(_owner, true);
        numKeyPairs++;
    }
    
     modifier onlyValidKeyPair(address _owner) {
        require(keyPairsContain(_owner), "Invalid key pair");
        _;
    }
    
    function keyPairsContain(address _owner) private view returns (bool) {
        for (uint256 i = 0; i < numKeyPairs; i++) {
            if (keyPairs[i].owner == _owner) {
                return true;
            }
        }
        return false;
    }
    
    function _validateTransaction(
        bytes32 _suggestedSignedHash,
        Transaction calldata _transaction
    ) internal onlyValidKeyPair(msg.sender) returns (bytes4 magic) {
       // the rest of the function as implemented in https://github.com/matter-labs/era-system-contracts/blob/main/contracts/DefaultAccount.sol
       
   ...
       
 ```
 
I believe you might want to invoke something similar to `addKeyPair` whenever a new session key is created and add the session public key in there as to allow for signature of transactions by this new key pair. 

Is this what you were thinking of?

**Comment by andybrierton** [comment]
In zkSync, the signature validation for transactions is primarily handled within the system contracts, specifically in the submitTx function. This function verifies the signature against the provided transaction data and ensures that it is valid for the specified sender address.

If you encounter the error message "Account validation returned invalid magic value," it typically indicates that the validation process within the system contract has failed. This could be due to various reasons such as incorrect signature data, mismatched sender address, or unexpected behavior during the validation process.

To troubleshoot this issue, you should review the implementation of the submitTx function in the zkSync system contracts, particularly the signature validation logic. Ensure that the provided signature matches the expected format and is being validated correctly against the sender address.

Additionally, double-check your Smart Account contract's implementation to confirm that the session key-pair management and signature generation are functioning as expected. Make sure that the session address is correctly recorded and used for signing transactions.

If the issue persists, consider reaching out to the zkSync development community for further assistance and debugging support. They may be able to provide insights or guidance on resolving the validation error.

**Comment by backvomeph** [comment]
The signature validation for transactions in zkSync is primarily conducted within the zkSync core protocol, outside the scope of individual smart contracts like the Smart Account contract. Ensure that transaction signatures adhere to zkSync's validation requirements and that the session key-pair setup is correctly implemented.

**Comment by kennycobbs** [comment]
It seems that the issue might be related to the validation process within the Zksync system contracts. Ensure that the signature validation for transactions signed by session keys is properly handled within the system contracts, possibly in methods other than those defined in the IAccount interface. Double-check the validation logic to ensure it accommodates transactions signed by session keys.

**Comment by postm4ltolin** [comment]
The user is trying to extend the IAccount interface to create a Smart Account contract for session key-pair management. They are encountering an error when executing a transaction with "from" set to the owner's address and signed with the session, and are receiving the error message "Account validation returned invalid magic value." They have checked the transaction execution flow and have not observed the methods from the IAccount interface within the Smart Account contract being executed. They are seeking guidance on where the signature validation is being performed other than the methods in the IAccount interface.

**Comment by andrewtotta** [comment]
The user is attempting to create a Smart Account contract for session key-pair management on zkSync's Testnet. They have encountered an issue where the validation of a transaction (signature) is not being conducted in the methods within the IAccount interface within the Smart Account contract. They have allowed the use of a session's private key to sign a transaction and encountered an error when executing a transaction with "from" set to the owner's address and signed with the session. The user is seeking guidance on where the validation of the transaction occurs and has not provided any relevant code.
